apt/minimization/orderedfeatureminimizer.py

```diff
@@ -0,0 +1,458 @@
+                                                                  right_id)
+        return X
+
+    @classmethod
+    def _transform_in_place(cls, dts, X, depths: Dict[str, int], numerical_features, categorical_features,
+                            feature_indices, generalizations_arrays):
+        for feature in numerical_features:
+            index = feature_indices[feature][0]
+            X[:, index] = cls._transform_numerical_feature(dts[feature].tree_, X[:, index],
+                                                           generalizations_arrays[feature], depths[feature], 0)
+
+        for feature in categorical_features:
+            indices = feature_indices[feature]
+            X[:, indices] = cls._transform_categorical_feature(dts[feature].tree_, X[:, indices],
+                                                               generalizations_arrays[feature], depths[feature], 0)
+
+    @classmethod
+    def populate_representative_median(cls, dt: Tree, X: np.ndarray, node_id, representative_values):
+        if X.size == 0:
+            return
+        threshold = dt.threshold[node_id]
+        representative_values[node_id] = np.median(X)
+        left_id = dt.children_left[node_id]
+        right_id = dt.children_right[node_id]
+        if left_id >= 0:
+            cls.populate_representative_median(dt, X[X <= threshold], left_id, representative_values)
+        if right_id >= 0:
+            cls.populate_representative_median(dt, X[X > threshold], right_id, representative_values)
+
+    @classmethod
+    def populate_representative_majority(cls, dt: Tree, X: np.ndarray, node_id, representative_values):
+        # since data is one hot encoded this line gets the majority index
+        if X.size == 0:
+            return
+        representative_values[node_id] = X.sum(axis=0).argmax()
+        split_feature = dt.feature[node_id]
+        threshold = dt.threshold[node_id]
+        left_id = dt.children_left[node_id]
+        right_id = dt.children_right[node_id]
+        # the threshold for encoded data is 0.5, therefore we take the column of the split feature and test threshold
+        # on it in order to split the data between the left and right child
+        if left_id >= 0:
+            cls.populate_representative_majority(dt, X[:, X[split_feature] <= threshold], left_id,
+                                                 representative_values)
+        if right_id >= 0:
+            cls.populate_representative_majority(dt, X[:, X[split_feature] > threshold], right_id,
+                                                 representative_values)
+
+    @classmethod
+    def _calculate_tree_depth(cls, dt, node_id=0):
+        if node_id < 0:
+            return 0
+        return 1 + max(cls._calculate_tree_depth(dt, dt.tree_.children_left[node_id]),
+                       cls._calculate_tree_depth(dt, dt.tree_.children_right[node_id]))
+
+    def fit(self, X: pd.DataFrame, y=None):
```

Comment:
Please add doctsring, and state the assumptions about the data (order of features etc.). Also, is it necessary that the input be a dataframe? Could we use the generic data wrappers (e.g., ArrayDataset) instead?

Reply:
Please add type of y